src/util/allocate_objects.cpp

```diff
@@ -0,0 +1,200 @@
+/*******************************************************************\
+
+Module:
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include "allocate_objects.h"
+
+#include <util/c_types.h>
+#include <util/fresh_symbol.h>
+#include <util/pointer_offset_size.h>
+
+/// Installs a new symbol in the symbol table, pushing the corresponding
+/// `symbolt` object to the field `symbols_created` and emits to `assignments`
+/// a new assignment of the form `<target_expr> := address-of(new_object)`. The
+/// `allocate_type` may differ from `target_expr.type()`, e.g. for target_expr
+/// having type `int*` and `allocate_type` being an `int[10]`.
+///
+/// \param assignments: The code block to add code to.
+/// \param target_expr: The expression which we are allocating a symbol for.
+/// \param allocate_type: Type of the object allocated
+/// \param alloc_type: Allocation type (global, local or dynamic)
```

Comment:
Yeah this is a bit tricky terminology-wise since as you mentioned the access scope, lifetime, and the location of the allocated memory are different things.

The way I see it is that the main thing `allocate_objectst` does is generate code that allocates objects (irrespective of where the memory allocated for those objects might be located physically). The code to allocate the objects does so by either declaring a local variable (local scope, automatic lifetime), global variable (global scope, static lifetime), or using `ALLOCATE()` (dynamic).

I was thinking of renaming the members of `allocation_typet` to `AUTOMATIC_LOCAL`, `STATIC_GLOBAL`, and `DYNAMIC`. Would you be ok with that?

Reply:
Would `lifetimet` be the most suitable name (instead of `allocation_typet`)? The suggested member names sound good to me, but mostly it's the use of "type" that makes me worried (hence suggesting `lifetimet`).

Reply:
Yeah, `liftetimet` sounds good to me. I'll change it to that.